guacamole-ext/src/main/java/org/apache/guacamole/properties/PrivateKeyGuacamoleProperty.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.properties;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.io.IOException;
+import java.lang.IllegalArgumentException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.security.spec.PKCS8EncodedKeySpec;
+import org.apache.guacamole.GuacamoleServerException;
+import org.apache.guacamole.environment.Environment;
+import org.apache.guacamole.environment.LocalEnvironment;
+
+/**
+ * A GuacamoleProperty whose value is derived from a private key file.
+ */
+public abstract class PrivateKeyGuacamoleProperty implements GuacamoleProperty<PrivateKey>  {
+
+    @Override
+    public PrivateKey parseValue(String value) throws GuacamoleServerException {
+
+        if (value == null || value.isEmpty())
+            return null;
+
+        try {
+
+            // Open and read the file specified in the configuration.
+            File keyFile = new File(value);
+            InputStream keyInput = new BufferedInputStream(new FileInputStream(keyFile));
+            int keyLength = (int) keyFile.length();
+            final byte[] keyBytes = new byte[keyLength];
```

Comment:
Beware that if you pre-allocate the `byte[]` like this, we'll need to bail out if the file size differs from what is expected. I don't think that's bad necessarily, so long as the case is handled.

The alternative would be to dynamically allocate the `byte[]`. An easy way would be to use something like a [`ByteArrayOutputStream`](https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html), reading/writing chunks in a loop.

Reply:
Okay, took a stab at a loop implementation of this...

Reply:
Looks better, however:

1. The duplicate `read()` calls are somewhat odd.
2. This loop implementation will spin indefinitely. When the buffer is filled, `keyBytes.length - totalBytesRead` becomes 0, in which case the call to `read()` will be asked to read 0 bytes and will always return 0, never signalling EOF.

It might be a lot easier to implement correctly if the byte array is allocated dynamically. That would also allow for cases where the file size is not known ahead of time, and remove the need to handle unexpected changes in file size.

Reply:
Well, I think I have a reasonable ByteArrayOutputStream solution, but, before I do that, any reason not to use Apache Commons IO?

https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#readFileToByteArray(java.io.File)

Reply:
> ... any reason not to use Apache Commons IO?

If you want to use Apache Commons IO, I'm not against it.

To me, the only reason would be to avoid bundling a library from which we're only going to use one function. We've removed dependencies in the past for similar reasons (see #37), but in that case there was a reasonable alternative already available within Java itself. If the alternative in the case of #37 was to write our own implementation entirely, I'm sure we would have kept the dependency around.

Reply:
Java 7 has a `readAllBytes` method in its `java.nio.file.Files` class:

http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#readAllBytes(java.nio.file.Path)

However, I thought I had seen some things in the pom.xml files that indicated that we try to maintain Java 6 compatibility, so I avoided that route.  If I'm mistaken about that or we're willing to put a Java 7 requirement on it, we could use the built-in and avoid both writing our own and pulling in a bundled library for a single method.

Reply:
While stopping support for Java 6 entirely would need some discussion, requiring Java 7 for a particular extension (such as the CAS auth) seems OK from my perspective.

Reply:
Weeeellllllllll...this particular change is for the newly-created `PrivateKeyGuacamoleProperty`, which is defined in `guacamole-ext`, not the CAS extension, so I think the implications are a little bit bigger than just CAS.

I'll just go ahead and go the `ByteArrayOutputStream` route for now, and then we can discuss on the dev list whether or not we want to push the compatibility up to Java 7 or even 8 (considering how long 8 has been out, and that 9 is out/almost out, and I think 10 is on the way).

Reply:
Oh.

Why is `PrivateKeyGuacamoleProperty` being added to the public extension API, rather than being internal to the CAS extension?

I'm not necessarily against adding a new property type at the extension API level, but PKCS8 + RSA is rather specific.

Reply:
I figured the property was generic enough that it might be useful to some other purpose in the future, so I put it in the guacamole-ext area instead of CAS.  I can move it if it makes sense at the moment - it can always be moved back up in the future if someone else finds it useful.

Reply:
Okay, so in the interest of getting this prepped more quickly for the 0.9.14-incubating release, I've moved the property to the CAS extension, and also implemented the the `ByteArrayOutputStream` solution instead of pulling in the extra package and changing the version requirement.